algopy_testing/src/algopy_testing/models/account.py

```diff
@@ -0,0 +1,157 @@
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Self, TypedDict, TypeVar
+
+import algosdk
+
+from algopy_testing.primitives.bytes import Bytes
+from algopy_testing.utils import as_string
+
+if TYPE_CHECKING:
+    import algopy
+
+
+T = TypeVar("T")
+
+
+class AccountFields(TypedDict, total=False):
+    balance: algopy.UInt64
+    min_balance: algopy.UInt64
+    auth_address: algopy.Account
+    total_num_uint: algopy.UInt64
+    total_num_byte_slice: algopy.Bytes
+    total_extra_app_pages: algopy.UInt64
+    total_apps_created: algopy.UInt64
+    total_apps_opted_in: algopy.UInt64
+    total_assets_created: algopy.UInt64
+    total_assets: algopy.UInt64
+    total_boxes: algopy.UInt64
+    total_box_bytes: algopy.UInt64
+
+
+@dataclass()
+class Account:
+    _public_key: str
+
+    def __init__(self, value: str | Bytes = algosdk.constants.ZERO_ADDRESS, /):
+        if not isinstance(value, (str | Bytes)):
+            raise TypeError("Invalid value for Account")
+        if isinstance(value, Bytes):
+            public_key = algosdk.encoding.encode_address(value)
+        else:
+            public_key = value
+        self._public_key = public_key
+
+    def is_opted_in(self, asset_or_app: algopy.Asset | algopy.Application, /) -> bool:
+        from algopy import Application, Asset
+
+        from algopy_testing import get_test_context
+
+        context = get_test_context()
+        opted_apps = context.account_data[self._public_key].opted_apps
+        opted_asset_balances = context.account_data[self._public_key].opted_asset_balances
+
+        if not context:
+            raise ValueError(
+                "Test context is not initialized! Use `with algopy_testing_context()` to access "
+                "the context manager."
+            )
+
+        if isinstance(asset_or_app, Asset):
+            return asset_or_app.id in opted_asset_balances
+        elif isinstance(asset_or_app, Application):
+            return asset_or_app.id in opted_apps
+
+        raise TypeError(
+            "Invalid `asset_or_app` argument type. Must be an `algopy.Asset` or "
+            "`algopy.Application` instance."
+        )
+
+    @classmethod
+    def from_bytes(cls, value: algopy.Bytes | bytes) -> Self:
+        from algopy import Bytes
+
+        if not isinstance(value, (Bytes | bytes)):
+            raise TypeError("Invalid value for Account")
+
+        value_len = value.length if isinstance(value, Bytes) else len(value)
+        if value_len != 32:
```

Comment:
Puya's implementation `from_bytes` does not do any validation beyond type, so to match the compiled behaviour this should not check the length or try to encode it